src/libraries/Microsoft.Extensions.Configuration/src/Config.cs

```diff
@@ -0,0 +1,376 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Configuration
+{
+    /// <summary>
+    /// Configuration is mutable configuration object. It is both an <see cref="IConfigurationBuilder"/> and an <see cref="IConfigurationRoot"/>.
+    /// As sources are added, it updates its current view of configuration. Once Build is called, configuration is frozen.
+    /// </summary>
+    public sealed class Config : IConfigurationBuilder, IConfigurationRoot, IDisposable
+    {
+        private readonly ConfigurationSources _sources;
+        private readonly ConfigurationBuilderProperties _properties;
+
+        private readonly object _providerLock = new();
+        private readonly List<IConfigurationProvider> _providers = new();
+        private readonly List<IDisposable> _changeTokenRegistrations = new();
+        private ConfigurationReloadToken _changeToken = new();
+
+        /// <summary>
+        /// Creates an empty mutable configuration object that is both an <see cref="IConfigurationBuilder"/> and an <see cref="IConfigurationRoot"/>.
+        /// </summary>
+        public Config()
+        {
+            _sources = new ConfigurationSources(this);
+            _properties = new ConfigurationBuilderProperties(this);
+
+            // Make sure there's some default storage since there are no default providers.
+            this.AddInMemoryCollection();
+
+            AddSource(_sources[0]);
+        }
+
+        /// <inheritdoc/>
+        public string this[string key]
+        {
+            get
+            {
+                lock (_providerLock)
+                {
+                    for (int i = _providers.Count - 1; i >= 0; i--)
```

Comment:
Now that this lives along side ConfigurationRoot, should we have an internal ConfigurationProviders class which can encapsulate the indexer so both the root and this class can share code?  Could also clean up the GetChildrenImplementation at the same time as a result

Reply:
Basically what I'm getting at is that it seems better to have one place that ConfigurationRoot and Config share the same code rather than having two copies of most of these methods that we'll need to keep in sync

Reply:
I shared some of this indexing code and called InternalConfigurationRootExtensions.GetChildrenImplementation() for GetChildren(). I don't think it makes sense to share much else. Is this good enough?